File: python/semantic_kernel/connectors/ai/google/google_ai/services/google_ai_chat_completion.py
# Copyright (c) Microsoft. All rights reserved.

import logging
import sys
from collections.abc import AsyncGenerator, Callable
from typing import TYPE_CHECKING, Any, ClassVar

if sys.version_info >= (3, 12):
    from typing import override  # pragma: no cover
else:
    from typing_extensions import override  # pragma: no cover

import google.generativeai as genai
from google.generativeai import GenerativeModel
from google.generativeai.protos import Candidate, Content
from google.generativeai.types import AsyncGenerateContentResponse, GenerateContentResponse, GenerationConfig
from pydantic import ValidationError

from semantic_kernel.connectors.ai.chat_completion_client_base import ChatCompletionClientBase
from semantic_kernel.connectors.ai.completion_usage import CompletionUsage
from semantic_kernel.connectors.ai.function_choice_behavior import FunctionChoiceType
from semantic_kernel.connectors.ai.google.google_ai.google_ai_prompt_execution_settings import (
    GoogleAIChatPromptExecutionSettings,
)
from semantic_kernel.connectors.ai.google.google_ai.google_ai_settings import GoogleAISettings
from semantic_kernel.connectors.ai.google.google_ai.services.google_ai_base import GoogleAIBase
from semantic_kernel.connectors.ai.google.google_ai.services.utils import (
    finish_reason_from_google_ai_to_semantic_kernel,
    format_assistant_message,
    format_tool_message,
    format_user_message,
    update_settings_from_function_choice_configuration,
)
from semantic_kernel.connectors.ai.google.shared_utils import (
    collapse_function_call_results_in_chat_history,
    filter_system_message,
    format_gemini_function_name_to_kernel_function_fully_qualified_name,
)
from semantic_kernel.contents.chat_history import ChatHistory
from semantic_kernel.contents.chat_message_content import CMC_ITEM_TYPES, ChatMessageContent
from semantic_kernel.contents.function_call_content import FunctionCallContent
from semantic_kernel.contents.streaming_chat_message_content import STREAMING_CMC_ITEM_TYPES as STREAMING_ITEM_TYPES
from semantic_kernel.contents.streaming_chat_message_content import StreamingChatMessageContent
from semantic_kernel.contents.streaming_text_content import StreamingTextContent
from semantic_kernel.contents.text_content import TextContent
from semantic_kernel.contents.utils.author_role import AuthorRole
from semantic_kernel.contents.utils.finish_reason import FinishReason
from semantic_kernel.exceptions.service_exceptions import (
    ServiceInitializationError,
    ServiceInvalidExecutionSettingsError,
)
from semantic_kernel.utils.telemetry.model_diagnostics.decorators import (
    trace_chat_completion,
    trace_streaming_chat_completion,
)

if TYPE_CHECKING:
    from semantic_kernel.connectors.ai.function_call_choice_configuration import FunctionCallChoiceConfiguration
    from semantic_kernel.connectors.ai.prompt_execution_settings import PromptExecutionSettings

logger: logging.Logger = logging.getLogger(__name__)


class GoogleAIChatCompletion(GoogleAIBase, ChatCompletionClientBase):
    """Google AI Chat Completion Client."""

    SUPPORTS_FUNCTION_CALLING: ClassVar[bool] = True

    def __init__(
        self,
        gemini_model_id: str | None = None,
        api_key: str | None = None,
        service_id: str | None = None,
        env_file_path: str | None = None,
        env_file_encoding: str | None = None,
    ) -> None:
        """Initialize the Google AI Chat Completion Client.

        If no arguments are provided, the service will attempt to load the settings from the environment.
        The following environment variables are used:
        - GOOGLE_AI_GEMINI_MODEL_ID
        - GOOGLE_AI_API_KEY

        Args:
            gemini_model_id (str | None): The Gemini model ID. (Optional)
            api_key (str | None): The API key. (Optional)
            service_id (str | None): The service ID. (Optional)
            env_file_path (str | None): The path to the .env file. (Optional)
            env_file_encoding (str | None): The encoding of the .env file. (Optional)

        Raises:
            ServiceInitializationError: If an error occurs during initialization.
        """
        try:
            google_ai_settings = GoogleAISettings(
                gemini_model_id=gemini_model_id,
                api_key=api_key,
                env_file_path=env_file_path,
                env_file_encoding=env_file_encoding,
            )
        except ValidationError as e:
            raise ServiceInitializationError(f"Failed to validate Google AI settings: {e}") from e
        if not google_ai_settings.gemini_model_id:
            raise ServiceInitializationError("The Google AI Gemini model ID is required.")

        super().__init__(
            ai_model_id=google_ai_settings.gemini_model_id,
            service_id=service_id or google_ai_settings.gemini_model_id,
            service_settings=google_ai_settings,
        )

    # region Overriding base class methods

    # Override from AIServiceClientBase
    @override
    def get_prompt_execution_settings_class(self) -> type["PromptExecutionSettings"]:
        return GoogleAIChatPromptExecutionSettings

    @override
    @trace_chat_completion(GoogleAIBase.MODEL_PROVIDER_NAME)
    async def _inner_get_chat_message_contents(
        self,
        chat_history: "ChatHistory",
        settings: "PromptExecutionSettings",
    ) -> list["ChatMessageContent"]:
        if not isinstance(settings, GoogleAIChatPromptExecutionSettings):
            settings = self.get_prompt_execution_settings_from_settings(settings)
        assert isinstance(settings, GoogleAIChatPromptExecutionSettings)  # nosec

        genai.configure(api_key=self.service_settings.api_key.get_secret_value())
        if not self.service_settings.gemini_model_id:
            raise ServiceInitializationError("The Google AI Gemini model ID is required.")
        model = GenerativeModel(
            model_name=self.service_settings.gemini_model_id,
            system_instruction=filter_system_message(chat_history),
        )

        collapse_function_call_results_in_chat_history(chat_history)

        response: AsyncGenerateContentResponse = await model.generate_content_async(
            contents=self._prepare_chat_history_for_request(chat_history),
            generation_config=GenerationConfig(**settings.prepare_settings_dict()),
            tools=settings.tools,
            tool_config=settings.tool_config,  # type: ignore
        )

        return [self._create_chat_message_content(response, candidate) for candidate in response.candidates]

    @override
    @trace_streaming_chat_completion(GoogleAIBase.MODEL_PROVIDER_NAME)
    async def _inner_get_streaming_chat_message_contents(
        self,
        chat_history: "ChatHistory",
        settings: "PromptExecutionSettings",
        function_invoke_attempt: int = 0,
    ) -> AsyncGenerator[list["StreamingChatMessageContent"], Any]:
        if not isinstance(settings, GoogleAIChatPromptExecutionSettings):
            settings = self.get_prompt_execution_settings_from_settings(settings)
        assert isinstance(settings, GoogleAIChatPromptExecutionSettings)  # nosec

        genai.configure(api_key=self.service_settings.api_key.get_secret_value())
        if not self.service_settings.gemini_model_id:
            raise ServiceInitializationError("The Google AI Gemini model ID is required.")
        model = GenerativeModel(
            model_name=self.service_settings.gemini_model_id,
            system_instruction=filter_system_message(chat_history),
        )

        collapse_function_call_results_in_chat_history(chat_history)

        response: AsyncGenerateContentResponse = await model.generate_content_async(
            contents=self._prepare_chat_history_for_request(chat_history),
            generation_config=GenerationConfig(**settings.prepare_settings_dict()),
            tools=settings.tools,
            tool_config=settings.tool_config,  # type: ignore
            stream=True,
        )

        async for chunk in response:
            yield [
                self._create_streaming_chat_message_content(chunk, candidate, function_invoke_attempt)
                for candidate in chunk.candidates
            ]

    @override
    def _verify_function_choice_settings(self, settings: "PromptExecutionSettings") -> None:
        if not isinstance(settings, GoogleAIChatPromptExecutionSettings):
            raise ServiceInvalidExecutionSettingsError("The settings must be an GoogleAIChatPromptExecutionSettings.")
        if settings.candidate_count is not None and settings.candidate_count > 1:
            raise ServiceInvalidExecutionSettingsError(
                "Auto-invocation of tool calls may only be used with a "
                "GoogleAIChatPromptExecutionSettings.candidate_count of 1."
            )

    @override
    def _update_function_choice_settings_callback(
        self,
    ) -> Callable[["FunctionCallChoiceConfiguration", "PromptExecutionSettings", FunctionChoiceType], None]:
        return update_settings_from_function_choice_configuration

    @override
    def _reset_function_choice_settings(self, settings: "PromptExecutionSettings") -> None:
        if hasattr(settings, "tool_config"):
            settings.tool_config = None
        if hasattr(settings, "tools"):
            settings.tools = None

    @override
    def _prepare_chat_history_for_request(
        self,
        chat_history: ChatHistory,
        role_key: str = "role",
        content_key: str = "content",
    ) -> list[Content]:
        chat_request_messages: list[Content] = []

        for message in chat_history.messages:
            if message.role == AuthorRole.SYSTEM:
                # Skip system messages since they are not part of the chat request.
                # System message will be provided as system_instruction in the model.
                continue
            if message.role == AuthorRole.USER:
                chat_request_messages.append(Content(role="user", parts=format_user_message(message)))
            elif message.role == AuthorRole.ASSISTANT:
                chat_request_messages.append(Content(role="model", parts=format_assistant_message(message)))
            elif message.role == AuthorRole.TOOL:
                chat_request_messages.append(Content(role="function", parts=format_tool_message(message)))

        return chat_request_messages

    # endregion

    # region Non-streaming

    def _create_chat_message_content(
        self, response: AsyncGenerateContentResponse, candidate: Candidate
    ) -> ChatMessageContent:
        """Create a chat message content object.

        Args:
            response: The response from the service.
            candidate: The candidate from the response.

        Returns:
            A chat message content object.
        """
        # Best effort conversion of finish reason. The raw value will be available in metadata.
        finish_reason: FinishReason | None = finish_reason_from_google_ai_to_semantic_kernel(candidate.finish_reason)
        response_metadata = self._get_metadata_from_response(response)
        response_metadata.update(self._get_metadata_from_candidate(candidate))

        items: list[CMC_ITEM_TYPES] = []
        for idx, part in enumerate(candidate.content.parts):
            if part.text:
                items.append(TextContent(text=part.text, inner_content=response, metadata=response_metadata))
            elif part.function_call:
                items.append(
                    FunctionCallContent(
                        id=f"{part.function_call.name}_{idx!s}",
                        name=format_gemini_function_name_to_kernel_function_fully_qualified_name(
                            part.function_call.name
                        ),
                        arguments={k: v for k, v in part.function_call.args.items()},
                    )
                )

        return ChatMessageContent(
            ai_model_id=self.ai_model_id,
            role=AuthorRole.ASSISTANT,
            items=items,
            inner_content=response,
            finish_reason=finish_reason,
            metadata=response_metadata,
        )

    # endregion

    # region Streaming

    def _create_streaming_chat_message_content(
        self,
        chunk: GenerateContentResponse,
        candidate: Candidate,
        function_invoke_attempt: int = 0,
    ) -> StreamingChatMessageContent:
        """Create a streaming chat message content object.

        Args:
            chunk: The response from the service.
            candidate: The candidate from the response.
            function_invoke_attempt: The function invoke attempt.

        Returns:
            A streaming chat message content object.
        """
        # Best effort conversion of finish reason. The raw value will be available in metadata.
        finish_reason: FinishReason | None = finish_reason_from_google_ai_to_semantic_kernel(candidate.finish_reason)
        response_metadata = self._get_metadata_from_response(chunk)
        response_metadata.update(self._get_metadata_from_candidate(candidate))

        items: list[STREAMING_ITEM_TYPES] = []
        for idx, part in enumerate(candidate.content.parts):
            if part.text:
                items.append(
                    StreamingTextContent(
                        choice_index=candidate.index,
                        text=part.text,
                        inner_content=chunk,
                        metadata=response_metadata,
                    )
                )
            elif part.function_call:
                items.append(
                    FunctionCallContent(
                        id=f"{part.function_call.name}_{idx!s}",
                        name=format_gemini_function_name_to_kernel_function_fully_qualified_name(
                            part.function_call.name
                        ),
                        arguments={k: v for k, v in part.function_call.args.items()},
                    )
                )

        return StreamingChatMessageContent(
            ai_model_id=self.ai_model_id,
            role=AuthorRole.ASSISTANT,
            choice_index=candidate.index,
            items=items,
            inner_content=chunk,
            finish_reason=finish_reason,
            metadata=response_metadata,
            function_invoke_attempt=function_invoke_attempt,
        )

    # endregion

    def _get_metadata_from_response(
        self, response: AsyncGenerateContentResponse | GenerateContentResponse
    ) -> dict[str, Any]:
        """Get metadata from the response.

        Args:
            response: The response from the service.

        Returns:
            A dictionary containing metadata.
        """
        return {
            "prompt_feedback": response.prompt_feedback,
            "usage": CompletionUsage(
                prompt_tokens=response.usage_metadata.prompt_token_count,
                completion_tokens=response.usage_metadata.candidates_token_count,
            ),
        }

    def _get_metadata_from_candidate(self, candidate: Candidate) -> dict[str, Any]:
        """Get metadata from the candidate.

        Args:
            candidate: The candidate from the response.

        Returns:
            A dictionary containing metadata.
        """
        return {
            "index": candidate.index,
            "finish_reason": candidate.finish_reason,
            "safety_ratings": candidate.safety_ratings,
            "token_count": candidate.token_count,
        }

File: nodejs/semantic-kernel/connectors/ai/google/google-ai/services/google-ai-chat-completion.ts
import { GoogleGenerativeAI } from '@google/generative-ai'
import { ChatHistory } from '../../../../../contents/chat-history'
import { ChatMessageContent } from '../../../../../contents/chat-message-content'
import { FunctionCallContent } from '../../../../../contents/function-call-content'
import { StreamingChatMessageContent } from '../../../../../contents/streaming-chat-message-content'
import { StreamingTextContent } from '../../../../../contents/streaming-text-content'
import { TextContent } from '../../../../../contents/text-content'
import { AuthorRole } from '../../../../../contents/utils/author-role'
import { FinishReason } from '../../../../../contents/utils/finish-reason'
import {
  ServiceInitializationError,
  ServiceInvalidExecutionSettingsError,
} from '../../../../../exceptions/service-exceptions'
import { PromptExecutionSettings } from '../../../../../services/ai-service-client-base'
import { ChatCompletionClientBase } from '../../../chat-completion-client-base'
import { CompletionUsage } from '../../../completion-usage'
import { FunctionCallChoiceConfiguration } from '../../../function-call-choice-configuration'
import { FunctionChoiceType } from '../../../function-choice-type'
import {
  collapseFunctionCallResultsInChatHistory,
  filterSystemMessage,
  formatGeminiFunctionNameToKernelFunctionFullyQualifiedName,
} from '../../shared-utils'
import { GoogleAIChatPromptExecutionSettings } from '../google-ai-prompt-execution-settings'
import { GoogleAISettings } from '../google-ai-settings'
import {
  finishReasonFromGoogleAIToSemanticKernel,
  formatAssistantMessage,
  formatToolMessage,
  formatUserMessage,
  Part,
  updateSettingsFromFunctionChoiceConfiguration,
} from './utils'

// Google AI SDK types
interface GenerateContentResponse {
  candidates: Candidate[]
  prompt_feedback?: any
  usage_metadata?: {
    prompt_token_count?: number
    candidates_token_count?: number
  }
}

interface Candidate {
  index?: number
  content?: {
    role?: string
    parts: Part[]
  }
  finish_reason?: string
  safety_ratings?: any[]
  token_count?: number
}

interface Content {
  role: string
  parts: Part[]
}

interface GenerateContentConfig {
  system_instruction?: string
  [key: string]: any
}

interface GoogleClient {
  aio: {
    models: {
      generate_content(params: {
        model: string
        contents: Content[]
        config: GenerateContentConfig
      }): Promise<GenerateContentResponse>

      generate_content_stream(params: {
        model: string
        contents: Content[]
        config: GenerateContentConfig
      }): Promise<AsyncIterable<GenerateContentResponse>>
    }
  }
}

/**
 * Google AI Chat Completion Client.
 * Provides chat completion capabilities using Google's Gemini models.
 */
export class GoogleAIChatCompletion extends ChatCompletionClientBase {
  static readonly MODEL_PROVIDER_NAME = 'googleai'
  static readonly SUPPORTS_FUNCTION_CALLING = true

  // Properties from GoogleAIBase pattern
  serviceSettings: GoogleAISettings
  client: any | null = null

  /**
   * Initialize the Google AI Chat Completion Client.
   *
   * If no arguments are provided, the service will attempt to load the settings from the environment.
   * The following environment variables are used:
   * - GOOGLE_AI_GEMINI_MODEL_ID
   * - GOOGLE_AI_API_KEY
   * - GOOGLE_AI_CLOUD_PROJECT_ID
   * - GOOGLE_AI_CLOUD_REGION
   * - GOOGLE_AI_USE_VERTEXAI
   *
   * @param params - Configuration parameters
   * @param params.geminiModelId - The Gemini model ID (Optional)
   * @param params.apiKey - The API key (Optional)
   * @param params.projectId - The Google Cloud project ID (Optional)
   * @param params.region - The Google Cloud region (Optional)
   * @param params.useVertexAI - Whether to use Vertex AI (Optional)
   * @param params.serviceId - The service ID (Optional)
   * @param params.client - The Google AI Client to use (Optional)
   * @throws {ServiceInitializationError} If validation fails or required settings are missing
   */
  constructor(
    params: {
      geminiModelId?: string | null
      apiKey?: string | null
      projectId?: string | null
      region?: string | null
      useVertexAI?: boolean | null
      serviceId?: string | null
      client?: GoogleClient | null
    } = {}
  ) {
    let googleAISettings: GoogleAISettings

    try {
      googleAISettings = new GoogleAISettings({
        geminiModelId: params.geminiModelId ?? undefined,
        apiKey: params.apiKey ?? undefined,
        cloudProjectId: params.projectId ?? undefined,
        cloudRegion: params.region ?? undefined,
        useVertexAI: params.useVertexAI ?? undefined,
      })
    } catch (error) {
      throw new ServiceInitializationError(
        `Failed to validate Google AI settings: ${error instanceof Error ? error.message : String(error)}`
      )
    }

    if (!googleAISettings.geminiModelId) {
      throw new ServiceInitializationError('The Google AI Gemini model ID is required.')
    }

    if (!params.client) {
      if (googleAISettings.useVertexAI && !googleAISettings.cloudProjectId) {
        throw new ServiceInitializationError('Project ID must be provided when useVertexAI is true.')
      }
      if (!googleAISettings.useVertexAI && !googleAISettings.apiKey) {
        throw new ServiceInitializationError('The API key is required when useVertexAI is false.')
      }
    }

    // Call parent constructor with aiModelId and serviceId
    super({
      aiModelId: googleAISettings.geminiModelId!,
      serviceId: params.serviceId || googleAISettings.geminiModelId,
    })

    // Initialize GoogleAI-specific properties
    this.serviceSettings = googleAISettings
    this.client = params.client ?? null
  }

  /**
   * Get the prompt execution settings class for this service.
   * @override
   */
  getPromptExecutionSettingsClass(): typeof GoogleAIChatPromptExecutionSettings {
    return GoogleAIChatPromptExecutionSettings
  }

  /**
   * Internal implementation for getting chat message contents.
   * This method is called by the base class's getChatMessageContents() method.
   *
   * @param chatHistory - The chat history to send
   * @param settings - Execution settings for the request
   * @returns Array of chat message content responses
   * @protected
   * @override
   */
  protected async _innerGetChatMessageContents(
    chatHistory: ChatHistory,
    settings: PromptExecutionSettings
  ): Promise<ChatMessageContent[]> {
    let executionSettings: GoogleAIChatPromptExecutionSettings

    if (!(settings instanceof GoogleAIChatPromptExecutionSettings)) {
      executionSettings = this.getPromptExecutionSettingsFromSettings(settings)
    } else {
      executionSettings = settings
    }
    // Type assertion for safety (matches Python implementation)
    if (!(executionSettings instanceof GoogleAIChatPromptExecutionSettings)) {
      throw new ServiceInitializationError('Failed to convert settings to GoogleAIChatPromptExecutionSettings')
    }

    if (!this.serviceSettings.geminiModelId) {
      throw new ServiceInitializationError('The Google AI Gemini model ID is required.')
    }

    collapseFunctionCallResultsInChatHistory(chatHistory)

    const generateContent = async (client: GoogleClient): Promise<GenerateContentResponse> => {
      return await client.aio.models.generate_content({
        model: this.serviceSettings.geminiModelId!,
        contents: this._prepareChatHistoryForRequest(chatHistory),
        config: {
          system_instruction: filterSystemMessage(chatHistory) ?? undefined,
          ...executionSettings.prepareSettingsDict(),
        },
      })
    }

    let response: GenerateContentResponse

    if (this.client) {
      response = await generateContent(this.client)
    } else if (this.serviceSettings.useVertexAI) {
      const client = this.createVertexAIClient()
      try {
        response = await generateContent(client)
      } finally {
        this.closeClient(client)
      }
    } else {
      const client = this.createAPIClient()
      try {
        response = await generateContent(client)
      } finally {
        this.closeClient(client)
      }
    }

    return response.candidates.map((candidate) => this._createChatMessageContent(response, candidate))
  }

  /**
   * Internal implementation for getting streaming chat message contents.
   * This method is called by the base class's getStreamingChatMessageContents() method.
   *
   * @param chatHistory - The chat history to send
   * @param settings - Execution settings for the request
   * @param functionInvokeAttempt - The function invoke attempt counter
   * @returns Async generator yielding arrays of streaming chat message content
   * @protected
   * @override
   */
  protected async *_innerGetStreamingChatMessageContents(
    chatHistory: ChatHistory,
    settings: PromptExecutionSettings,
    functionInvokeAttempt: number = 0
  ): AsyncGenerator<StreamingChatMessageContent[], void> {
    let executionSettings: GoogleAIChatPromptExecutionSettings

    if (!(settings instanceof GoogleAIChatPromptExecutionSettings)) {
      executionSettings = this.getPromptExecutionSettingsFromSettings(settings)
    } else {
      executionSettings = settings
    }
    // Type assertion for safety (matches Python implementation)
    if (!(executionSettings instanceof GoogleAIChatPromptExecutionSettings)) {
      throw new ServiceInitializationError('Failed to convert settings to GoogleAIChatPromptExecutionSettings')
    }

    if (!this.serviceSettings.geminiModelId) {
      throw new ServiceInitializationError('The Google AI Gemini model ID is required.')
    }

    collapseFunctionCallResultsInChatHistory(chatHistory)

    const generateContentStream = async function* (
      this: GoogleAIChatCompletion,
      client: GoogleClient
    ): AsyncGenerator<GenerateContentResponse, void> {
      const stream = await client.aio.models.generate_content_stream({
        model: this.serviceSettings.geminiModelId!,
        contents: this._prepareChatHistoryForRequest(chatHistory),
        config: {
          system_instruction: filterSystemMessage(chatHistory) ?? undefined,
          ...executionSettings.prepareSettingsDict(),
        },
      })

      for await (const chunk of stream) {
        yield chunk
      }
    }.bind(this)

    if (this.client) {
      for await (const chunk of generateContentStream(this.client)) {
        yield chunk.candidates.map((candidate) =>
          this._createStreamingChatMessageContent(chunk, candidate, functionInvokeAttempt)
        )
      }
    } else if (this.serviceSettings.useVertexAI) {
      const client = this.createVertexAIClient()
      try {
        for await (const chunk of generateContentStream(client)) {
          yield chunk.candidates.map((candidate) =>
            this._createStreamingChatMessageContent(chunk, candidate, functionInvokeAttempt)
          )
        }
      } finally {
        this.closeClient(client)
      }
    } else {
      const client = this.createAPIClient()
      try {
        for await (const chunk of generateContentStream(client)) {
          yield chunk.candidates.map((candidate) =>
            this._createStreamingChatMessageContent(chunk, candidate, functionInvokeAttempt)
          )
        }
      } finally {
        this.closeClient(client)
      }
    }
  }

  /**
   * Verify function choice settings are valid.
   *
   * @param settings - The settings to verify
   * @protected
   * @override
   */
  protected _verifyFunctionChoiceSettings(settings: PromptExecutionSettings): void {
    if (!(settings instanceof GoogleAIChatPromptExecutionSettings)) {
      throw new ServiceInvalidExecutionSettingsError('The settings must be a GoogleAIChatPromptExecutionSettings.')
    }
    if (settings.candidateCount !== null && settings.candidateCount !== undefined && settings.candidateCount > 1) {
      throw new ServiceInvalidExecutionSettingsError(
        'Auto-invocation of tool calls may only be used with a GoogleAIChatPromptExecutionSettings.candidateCount of 1.'
      )
    }
  }

  /**
   * Get the callback for updating function choice settings.
   *
   * @returns The callback function
   * @protected
   * @override
   */
  protected _updateFunctionChoiceSettingsCallback(): (
    configuration: FunctionCallChoiceConfiguration,
    settings: PromptExecutionSettings,
    choiceType: FunctionChoiceType
  ) => void {
    return (
      configuration: FunctionCallChoiceConfiguration,
      settings: PromptExecutionSettings,
      choiceType: FunctionChoiceType
    ) => {
      updateSettingsFromFunctionChoiceConfiguration(configuration, settings, choiceType)
    }
  }

  /**
   * Reset function choice settings.
   *
   * @param settings - The settings to reset
   * @protected
   * @override
   */
  protected _resetFunctionChoiceSettings(settings: PromptExecutionSettings): void {
    if ('toolConfig' in settings) {
      ;(settings as any).toolConfig = null
    }
    if ('tools' in settings) {
      ;(settings as any).tools = null
    }
  }

  /**
   * Prepare chat history for request.
   *
   * @param chatHistory - The chat history to prepare
   * @param roleKey - The role key (unused, for compatibility)
   * @param contentKey - The content key (unused, for compatibility)
   * @returns The prepared chat history
   * @protected
   * @override
   */
  protected _prepareChatHistoryForRequest(
    chatHistory: ChatHistory,
    _roleKey: string = 'role',
    _contentKey: string = 'content'
  ): Content[] {
    const chatRequestMessages: Content[] = []

    for (const message of chatHistory.messages) {
      if (message.role === AuthorRole.SYSTEM) {
        // Skip system messages since they are not part of the chat request.
        // System message will be provided as system_instruction in the config.
        continue
      }
      if (message.role === AuthorRole.USER) {
        chatRequestMessages.push({ role: 'user', parts: formatUserMessage(message) })
      } else if (message.role === AuthorRole.ASSISTANT) {
        chatRequestMessages.push({ role: 'model', parts: formatAssistantMessage(message) })
      } else if (message.role === AuthorRole.TOOL) {
        chatRequestMessages.push({ role: 'function', parts: formatToolMessage(message) })
      }
    }

    return chatRequestMessages
  }

  // #region Helper methods

  /**
   * Create a chat message content object from a response and candidate.
   *
   * @param response - The response from the service
   * @param candidate - The candidate from the response
   * @returns A chat message content object
   * @private
   */
  private _createChatMessageContent(response: GenerateContentResponse, candidate: Candidate): ChatMessageContent {
    // Best effort conversion of finish reason. The raw value will be available in metadata.
    const finishReason: FinishReason | null = finishReasonFromGoogleAIToSemanticKernel(candidate.finish_reason)
    const responseMetadata = {
      ...this._getMetadataFromResponse(response),
      ...this._getMetadataFromCandidate(candidate),
    }

    const items: Array<TextContent | FunctionCallContent> = []

    if (candidate.content && candidate.content.parts) {
      for (let idx = 0; idx < candidate.content.parts.length; idx++) {
        const part = candidate.content.parts[idx]
        if (part.text) {
          items.push(
            new TextContent({
              text: part.text,
              innerContent: response,
              metadata: responseMetadata,
            })
          )
        } else if (part.functionCall) {
          items.push(
            new FunctionCallContent({
              id: `${part.functionCall.name}_${idx}`,
              name: formatGeminiFunctionNameToKernelFunctionFullyQualifiedName(part.functionCall.name),
              arguments: { ...part.functionCall.args },
            })
          )
        }
      }
    }

    return new ChatMessageContent({
      aiModelId: this.aiModelId,
      role: AuthorRole.ASSISTANT,
      items,
      innerContent: response,
      finishReason: finishReason ?? undefined,
      metadata: responseMetadata,
    })
  }

  /**
   * Create a streaming chat message content object from a chunk and candidate.
   *
   * @param chunk - The response chunk from the service
   * @param candidate - The candidate from the chunk
   * @param functionInvokeAttempt - The function invoke attempt counter
   * @returns A streaming chat message content object
   * @private
   */
  private _createStreamingChatMessageContent(
    chunk: GenerateContentResponse,
    candidate: Candidate,
    functionInvokeAttempt: number = 0
  ): StreamingChatMessageContent {
    // Best effort conversion of finish reason. The raw value will be available in metadata.
    const finishReason: FinishReason | null = finishReasonFromGoogleAIToSemanticKernel(candidate.finish_reason)
    const responseMetadata = {
      ...this._getMetadataFromResponse(chunk),
      ...this._getMetadataFromCandidate(candidate),
    }

    const items: Array<StreamingTextContent | FunctionCallContent> = []

    if (candidate.content && candidate.content.parts) {
      for (let idx = 0; idx < candidate.content.parts.length; idx++) {
        const part = candidate.content.parts[idx]
        if (part.text) {
          items.push(
            new StreamingTextContent({
              choiceIndex: candidate.index ?? 0,
              text: part.text,
              innerContent: chunk,
              metadata: responseMetadata,
            })
          )
        } else if (part.functionCall) {
          items.push(
            new FunctionCallContent({
              id: `${part.functionCall.name}_${idx}`,
              name: formatGeminiFunctionNameToKernelFunctionFullyQualifiedName(part.functionCall.name),
              arguments: { ...part.functionCall.args },
            })
          )
        }
      }
    }

    return new StreamingChatMessageContent({
      aiModelId: this.aiModelId,
      role: AuthorRole.ASSISTANT,
      choiceIndex: candidate.index ?? 0,
      items,
      innerContent: chunk,
      finishReason: finishReason ?? undefined,
      metadata: responseMetadata,
      functionInvokeAttempt,
    })
  }

  /**
   * Extract metadata from a response object.
   *
   * @param response - The response from the service
   * @returns A dictionary containing metadata
   * @private
   */
  private _getMetadataFromResponse(response: GenerateContentResponse): Record<string, any> {
    return {
      prompt_feedback: response.prompt_feedback,
      usage: new CompletionUsage({
        promptTokens: response.usage_metadata ? (response.usage_metadata.prompt_token_count ?? undefined) : undefined,
        completionTokens: response.usage_metadata
          ? (response.usage_metadata.candidates_token_count ?? undefined)
          : undefined,
      }),
    }
  }

  /**
   * Extract metadata from a candidate object.
   *
   * @param candidate - The candidate from the response
   * @returns A dictionary containing metadata
   * @private
   */
  private _getMetadataFromCandidate(candidate: Candidate): Record<string, any> {
    return {
      index: candidate.index,
      finish_reason: candidate.finish_reason,
      safety_ratings: candidate.safety_ratings,
      token_count: candidate.token_count,
    }
  }

  /**
   * Helper method to create a Vertex AI client.
   * Implements the client creation pattern from GoogleAIBase.
   *
   * Creates a client using Vertex AI with the configured project ID and region.
   * Equivalent to Python's: Client(vertexai=True, project=..., location=...)
   *
   * @private
   */
  private createVertexAIClient(): GoogleClient {
    // This would use the Google GenAI SDK's Client class:
    // const { Client } = require('@google/generative-ai') or similar
    // return new Client({
    //   vertexai: true,
    //   project: this.serviceSettings.cloudProjectId,
    //   location: this.serviceSettings.cloudRegion,
    // })
    throw new Error(
      'Vertex AI client creation requires the Google GenAI SDK. ' +
        'Install the package and implement client creation using: ' +
        'new Client({ vertexai: true, project: cloudProjectId, location: cloudRegion })'
    )
  }

  /**
   * Helper method to create an API key-based client.
   * Implements the client creation pattern from GoogleAIBase.
   *
   * Creates a client using an API key for authentication.
   * Equivalent to Python's: Client(api_key=api_key)
   *
   * @private
   */
  private createAPIClient(): GoogleClient {
    if (!this.serviceSettings.apiKey) {
      throw new ServiceInitializationError('API key is required to create a client.')
    }

    const genAI = new GoogleGenerativeAI(this.serviceSettings.apiKey)
    const model = genAI.getGenerativeModel({ model: this.serviceSettings.geminiModelId! })

    // Wrap the Google Generative AI client to match our GoogleClient interface
    return {
      aio: {
        models: {
          generate_content: async (params: {
            model: string
            contents: Content[]
            config: GenerateContentConfig
          }): Promise<GenerateContentResponse> => {
            const result = await model.generateContent({
              contents: params.contents,
              systemInstruction: params.config.system_instruction,
              generationConfig: params.config,
            } as any)
            const response = result.response
            return {
              candidates:
                response.candidates?.map((candidate) => ({
                  index: candidate.index,
                  content: candidate.content,
                  finish_reason: candidate.finishReason,
                  safety_ratings: candidate.safetyRatings,
                })) || [],
              prompt_feedback: response.promptFeedback,
              usage_metadata: {
                prompt_token_count: response.usageMetadata?.promptTokenCount,
                candidates_token_count: response.usageMetadata?.candidatesTokenCount,
              },
            }
          },
          generate_content_stream: async (params: {
            model: string
            contents: Content[]
            config: GenerateContentConfig
          }): Promise<AsyncIterable<GenerateContentResponse>> => {
            const result = await model.generateContentStream({
              contents: params.contents,
              systemInstruction: params.config.system_instruction,
              generationConfig: params.config,
            } as any)

            async function* streamWrapper(): AsyncGenerator<GenerateContentResponse, void> {
              for await (const chunk of result.stream) {
                yield {
                  candidates:
                    chunk.candidates?.map((candidate) => ({
                      index: candidate.index,
                      content: candidate.content,
                      finish_reason: candidate.finishReason,
                      safety_ratings: candidate.safetyRatings,
                    })) || [],
                  prompt_feedback: chunk.promptFeedback,
                  usage_metadata: {
                    prompt_token_count: chunk.usageMetadata?.promptTokenCount,
                    candidates_token_count: chunk.usageMetadata?.candidatesTokenCount,
                  },
                }
              }
            }

            return streamWrapper()
          },
        },
      },
    }
  }

  /**
   * Helper method to close/cleanup a client.
   *
   * @param _client - The client to clean up
   * @private
   */
  private closeClient(_client: GoogleClient): void {
    // Implement cleanup if needed
    // Python uses context managers (with statement), TypeScript might need explicit cleanup
  }

  /**
   * Convert generic settings to Google AI specific settings.
   * Overrides the base class method to provide Google AI specific conversion.
   *
   * @param settings - Generic prompt execution settings
   * @returns Google AI specific settings
   * @override
   */
  getPromptExecutionSettingsFromSettings(settings: PromptExecutionSettings): GoogleAIChatPromptExecutionSettings {
    // Convert or validate settings to Google AI specific format
    if (settings instanceof GoogleAIChatPromptExecutionSettings) {
      return settings
    }

    // Create new Google AI settings from generic settings
    return new GoogleAIChatPromptExecutionSettings()
  }

  // #endregion
}
